File: app/src/main/java/com/zacharee1/systemuituner/misc/QSDragAdapter.java
package com.zacharee1.systemuituner.misc;

import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.pm.PackageManager;
import android.content.res.Resources;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.os.Build;
import android.provider.Settings;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.zacharee1.systemuituner.R;
import com.zacharee1.systemuituner.util.SettingsUtils;

import java.util.ArrayList;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class QSDragAdapter extends RecyclerView.Adapter<QSDragAdapter.QSViewHolder> {
    public ArrayList<QSTile> mTiles = new ArrayList<>();

    public ArrayList<QSTile> mAvailableTiles = new ArrayList<>();

    private Context mContext;

    public QSDragAdapter(Context context) {
        mContext = context;
        parseTileList();
    }

    public void parseTileList() {
        String tiles = Settings.Secure.getString(mContext.getContentResolver(), "sysui_qs_tiles");

        if (tiles == null) {
            tiles = getDefaultTileOrder();
        }

        String[] tileArray = tiles.split("[,]");

        ArrayList<QSTile> tempTiles = new ArrayList<>();

        for (String tile : tileArray) {
            tempTiles.add(new QSTile(tile, mContext));
        }

        mTiles.clear();
        mTiles.addAll(tempTiles);

        refreshAvailableTiles();
    }

    public void refreshAvailableTiles() {
        mAvailableTiles.clear();
        for (QSTile tile : getDefaultTiles()) {
            boolean hasTile = false;

            for (QSTile tile1 : mTiles) {
                if (tile1.key.equals(tile.key)) hasTile = true;
            }

            if (!hasTile) {
                mAvailableTiles.add(tile);
            }
        }
    }

    private ArrayList<QSTile> getDefaultTiles() {
        String order = getDefaultTileOrder();
        String[] array = order.split("[,]");

        ArrayList<QSTile> ret = new ArrayList<>();

        for (String tile : array) {
            ret.add(new QSTile(tile, mContext));
        }

        return ret;
    }

    private String getDefaultTileOrder() {
        PackageManager pm = mContext.getPackageManager();

        try {
            Resources resources = pm.getResourcesForApplication("com.android.systemui");
            int id = resources.getIdentifier("quick_settings_tiles_default", "string", "com.android.systemui");

            return resources.getString(id);
        } catch (Exception e) {
            return "";
        }
    }

    public void addTile(QSTile tile) {
        mTiles.add(tile);
        notifyDataSetChanged();

        setOrder(mTiles);
        refreshAvailableTiles();
    }

    public void setOrder(ArrayList<QSDragAdapter.QSTile> tiles) {
        ArrayList<String> keys = new ArrayList<>();

        for (QSTile tile : tiles) {
            keys.add(tile.key);
        }

        String tileString = TextUtils.join(",", keys);

        SettingsUtils.writeSecure(mContext, "sysui_qs_tiles", tileString);
    }

    @Override
    public QSViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.qs_tile_layout, parent, false);
        return new QSViewHolder(view);
    }

    @Override
    public void onBindViewHolder(final QSViewHolder holder, int position) {
        holder.setTitle(mTiles.get(holder.getAdapterPosition()).title);
        holder.setIcon(mTiles.get(holder.getAdapterPosition()).icon);
        holder.setCloseListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                new AlertDialog.Builder(mContext)
                        .setTitle(R.string.removing_tile)
                        .setMessage(String.format(holder.getContext().getResources().getString(R.string.remove_tile), mTiles.get(holder.getAdapterPosition()).title))
                        .setPositiveButton(R.string.yes, new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialogInterface, int i) {
                                mTiles.remove(holder.getAdapterPosition());
                                setOrder(mTiles);
                                notifyItemRemoved(holder.getAdapterPosition());
                            }
                        })
                        .setNegativeButton(R.string.no, null)
                        .show();
            }
        });
    }

    @Override
    public int getItemCount() {
        return mTiles.size();
    }

    public static class QSViewHolder extends RecyclerView.ViewHolder {
        View mView;

        public QSViewHolder(View view) {
            super(view);
            mView = view;

            if (Build.VERSION.SDK_INT > Build.VERSION_CODES.M) {
                mView.findViewById(R.id.close_button).setVisibility(View.GONE);
            }
        }

        public void setTitle(String title) {
            TextView textView = mView.findViewById(R.id.textView);
            textView.setText(title);
        }

        public void setIcon(Drawable icon) {
            ImageView imageView = mView.findViewById(R.id.imageView);
            imageView.setImageDrawable(icon);
        }

        public void setCloseListener(View.OnClickListener listener) {
            mView.findViewById(R.id.close_button).setOnClickListener(listener);
        }

        public Context getContext() {
            return mView.getContext();
        }
    }

    public static class QSTile {
        public String title;
        public Drawable icon;
        public String key;

        public TileParser mParser;

        public QSTile(String key, Context context) {
            this.key = key;
            mParser = new TileParser(key, context);

            this.title = mParser.title;
            this.icon = mParser.icon;
        }
    }

    public static class TileParser {
        public String key;
        public Drawable icon;
        public String title;

        private Context mContext;

        public TileParser(String key, Context context) {
            this.key = key;
            mContext = context;

            parseKey();
        }

        private void parseKey() {
            if (key.toLowerCase().contains("intent(")) {
                parseIntent();
            } else if (key.toLowerCase().contains("custom(")) {
                parseCustom();
            } else {
                parseStandard();
            }
        }

        private void parseIntent() {
            Drawable drawable = mContext.getResources().getDrawable(R.drawable.ic_android_black_24dp, null);
            drawable.setColorFilter(Color.WHITE, PorterDuff.Mode.SRC_ATOP);

            icon = drawable;

            Pattern p = Pattern.compile("\\((.*?)\\)");
            Matcher m = p.matcher(key);

            String title = "";

            while (!m.hitEnd()) {
                if (m.find()) title = m.group();
            }

            this.title = title.replace("(", "").replace(")", "");
        }

        private void parseCustom() {
            Pattern p = Pattern.compile("\\((.*?)\\)");
            Matcher m = p.matcher(key);

            String name = "";

            while (!m.hitEnd()) {
                if (m.find()) name = m.group();
            }

            name = name.replace("(", "").replace(")", "");

            String packageName = name.split("[/]")[0];
            String component = name.split("[/]")[1];

            try {
                icon = mContext.getPackageManager().getApplicationIcon(packageName);
            } catch (Exception e) {}

            try {
                String[] split = component.split("[.]");
                title = split[split.length - 1];
            } catch (Exception e) {}
        }

        private void parseStandard() {
            title = key.toLowerCase();

            int iconRes = R.drawable.ic_android_black_24dp;

            switch (key.toLowerCase()) {
                case "wifi":
                    iconRes = R.drawable.ic_signal_wifi_4_bar_black_24dp;
                    break;
                case "bluetooth":
                case "bt":
                    iconRes = R.drawable.ic_bluetooth_black_24dp;
                    break;
                case "color_inversion":
                case "inversion":
                    iconRes = R.drawable.ic_invert_colors_black_24dp;
                    break;
                case "cell":
                    iconRes = R.drawable.ic_signal_cellular_4_bar_black_24dp;
                    break;
                case "do_not_disturb":
                case "dnd":
                    iconRes = R.drawable.ic_do_not_disturb_on_black_24dp;
                    break;
                case "airplane":
                    iconRes = R.drawable.ic_airplanemode_active_black_24dp;
                    break;
                case "cast":
                    iconRes = R.drawable.ic_cast_black_24dp;
                    break;
                case "location":
                    iconRes = R.drawable.ic_location_on_black_24dp;
                    break;
                case "rotation":
                    iconRes = R.drawable.ic_screen_rotation_black_24dp;
                    break;
                case "flashlight":
                    iconRes = R.drawable.ic_highlight_black_24dp;
                    break;
                case "hotspot":
                    iconRes = R.drawable.ic_wifi_tethering_black_24dp;
                    break;
                case "battery":
                    iconRes = R.drawable.ic_battery_full_black_24dp;
                    break;
                case "sound":
                    iconRes = R.drawable.ic_volume_up_black_24dp;
                    break;
                case "sync":
                    iconRes = R.drawable.ic_sync_black_24dp;
                    break;
                case "nfc":
                    iconRes = R.drawable.ic_nfc_black_24dp;
                    break;
                case "data":
                    iconRes = R.drawable.ic_data_usage_black_24dp;
                    break;
            }

            Drawable drawable = mContext.getResources().getDrawable(iconRes, null).getCurrent().mutate();
            drawable.setColorFilter(Color.WHITE, PorterDuff.Mode.SRC_ATOP);

            icon = drawable;
        }
    }
}

File: app/src/main/java/com/zacharee1/systemuituner/misc/QSDragAdapter.kt
package com.zacharee1.systemuituner.misc

import android.app.AlertDialog
import android.content.Context
import android.graphics.Color
import android.graphics.PorterDuff
import android.graphics.drawable.Drawable
import android.os.Build
import android.provider.Settings
import android.support.v7.widget.RecyclerView
import android.text.TextUtils
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import com.zacharee1.systemuituner.R
import com.zacharee1.systemuituner.util.SettingsUtils
import java.util.*
import java.util.regex.Pattern

class QSDragAdapter(private val mContext: Context) : RecyclerView.Adapter<QSDragAdapter.QSViewHolder>() {
    var mTiles = ArrayList<QSTile>()

    var mAvailableTiles = ArrayList<QSTile>()

    private val defaultTiles: ArrayList<QSTile>
        get() {
            val order = defaultTileOrder
            val array = order.split(",")

            return array.mapTo(ArrayList()) { QSTile(it, mContext) }
        }

    private val defaultTileOrder: String
        get() {
            val pm = mContext.packageManager

            return try {
                val resources = pm.getResourcesForApplication("com.android.systemui")
                val id = resources.getIdentifier("quick_settings_tiles_default", "string", "com.android.systemui")

                resources.getString(id)
            } catch (e: Exception) {
                ""
            }

        }

    init {
        parseTileList()
    }

    fun parseTileList() {
        var tiles: String? = Settings.Secure.getString(mContext.contentResolver, "sysui_qs_tiles")

        if (tiles == null) {
            tiles = defaultTileOrder
        }

        val tileArray = tiles.split(",")

        val tempTiles = tileArray.map { QSTile(it, mContext) }

        mTiles.clear()
        mTiles.addAll(tempTiles)

        refreshAvailableTiles()
    }

    private fun refreshAvailableTiles() {
        mAvailableTiles.clear()
        for (tile in defaultTiles) {
            val hasTile = mTiles.any { it.key == tile.key }

            if (!hasTile) {
                mAvailableTiles.add(tile)
            }
        }
    }

    fun addTile(tile: QSTile) {
        mTiles.add(tile)
        notifyDataSetChanged()

        setOrder(mTiles)
        refreshAvailableTiles()
    }

    fun setOrder(tiles: ArrayList<QSDragAdapter.QSTile>) {
        val keys = tiles.map { it.key }

        val tileString = TextUtils.join(",", keys)

        SettingsUtils.writeSecure(mContext, "sysui_qs_tiles", tileString)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): QSViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.qs_tile_layout, parent, false)
        return QSViewHolder(view)
    }

    override fun onBindViewHolder(holder: QSViewHolder, position: Int) {
        holder.setTitle(mTiles[holder.adapterPosition].title)
        holder.setIcon(mTiles[holder.adapterPosition].icon)
        holder.setCloseListener(View.OnClickListener {
            AlertDialog.Builder(mContext)
                    .setTitle(R.string.removing_tile)
                    .setMessage(String.format(holder.context.resources.getString(R.string.remove_tile), mTiles[holder.adapterPosition].title))
                    .setPositiveButton(R.string.yes, { _, _ ->
                        mTiles.removeAt(holder.adapterPosition)
                        setOrder(mTiles)
                        notifyItemRemoved(holder.adapterPosition)
                    })
                    .setNegativeButton(R.string.no, null)
                    .show()
        })
    }

    override fun getItemCount(): Int {
        return mTiles.size
    }

    class QSViewHolder(private var mView: View) : RecyclerView.ViewHolder(mView) {

        val context: Context
            get() = mView.context

        init {

            if (Build.VERSION.SDK_INT > Build.VERSION_CODES.M) {
                mView.findViewById<View>(R.id.close_button).visibility = View.GONE
            }
        }

        fun setTitle(title: String) {
            val textView = mView.findViewById<TextView>(R.id.textView)
            textView.text = title
        }

        fun setIcon(icon: Drawable) {
            val imageView = mView.findViewById<ImageView>(R.id.imageView)
            imageView.setImageDrawable(icon)
        }

        fun setCloseListener(listener: View.OnClickListener) {
            mView.findViewById<View>(R.id.close_button).setOnClickListener(listener)
        }
    }

    class QSTile(var key: String, context: Context) {
        var title: String
        var icon: Drawable

        private var mParser: TileParser = TileParser(key, context)

        init {

            this.title = mParser.title
            this.icon = mParser.icon
        }
    }

    class TileParser(var key: String, private val mContext: Context) {
        lateinit var icon: Drawable
        lateinit var title: String

        init {

            parseKey()
        }

        private fun parseKey() {
            when {
                key.toLowerCase().contains("intent(") -> parseIntent()
                key.toLowerCase().contains("custom(") -> parseCustom()
                else -> parseStandard()
            }
        }

        private fun parseIntent() {
            val drawable = mContext.resources.getDrawable(R.drawable.ic_android_black_24dp, null)
            drawable.setColorFilter(Color.WHITE, PorterDuff.Mode.SRC_ATOP)

            icon = drawable

            val p = Pattern.compile("\\((.*?)\\)")
            val m = p.matcher(key)

            var title = ""

            while (!m.hitEnd()) {
                if (m.find()) title = m.group()
            }

            this.title = title.replace("(", "").replace(")", "")
        }

        private fun parseCustom() {
            val p = Pattern.compile("\\((.*?)\\)")
            val m = p.matcher(key)

            var name = ""

            while (!m.hitEnd()) {
                if (m.find()) name = m.group()
            }

            name = name.replace("(", "").replace(")", "")

            val packageName = name.split("/")[0]
            val component = name.split("/")[1]

            try {
                icon = mContext.packageManager.getApplicationIcon(packageName)
            } catch (e: Exception) {
            }

            try {
                val split = component.split(".")
                title = split[split.size - 1]
            } catch (e: Exception) {
            }

        }

        private fun parseStandard() {
            title = key.toLowerCase()

            var iconRes = R.drawable.ic_android_black_24dp

            when (key.toLowerCase()) {
                "wifi" -> iconRes = R.drawable.ic_signal_wifi_4_bar_black_24dp
                "bluetooth", "bt" -> iconRes = R.drawable.ic_bluetooth_black_24dp
                "color_inversion", "inversion" -> iconRes = R.drawable.ic_invert_colors_black_24dp
                "cell" -> iconRes = R.drawable.ic_signal_cellular_4_bar_black_24dp
                "do_not_disturb", "dnd" -> iconRes = R.drawable.ic_do_not_disturb_on_black_24dp
                "airplane" -> iconRes = R.drawable.ic_airplanemode_active_black_24dp
                "cast" -> iconRes = R.drawable.ic_cast_black_24dp
                "location" -> iconRes = R.drawable.ic_location_on_black_24dp
                "rotation" -> iconRes = R.drawable.ic_screen_rotation_black_24dp
                "flashlight" -> iconRes = R.drawable.ic_highlight_black_24dp
                "hotspot" -> iconRes = R.drawable.ic_wifi_tethering_black_24dp
                "battery" -> iconRes = R.drawable.ic_battery_full_black_24dp
                "sound" -> iconRes = R.drawable.ic_volume_up_black_24dp
                "sync" -> iconRes = R.drawable.ic_sync_black_24dp
                "nfc" -> iconRes = R.drawable.ic_nfc_black_24dp
                "data" -> iconRes = R.drawable.ic_data_usage_black_24dp
            }

            val drawable = mContext.resources.getDrawable(iconRes, null).current.mutate()
            drawable.setColorFilter(Color.WHITE, PorterDuff.Mode.SRC_ATOP)

            icon = drawable
        }
    }
}
